presto-base-arrow-flight/src/test/java/com/facebook/plugin/arrow/TestingArrowMetadata.java

```diff
@@ -0,0 +1,157 @@
+    {
+        return new TestingArrowFlightRequest(config, schema, nodeManager.getWorkerNodes().size(), testconfig);
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        List<String> listSchemas = extractSchemaAndTableData(Optional.empty(), session);
+        List<String> names = new ArrayList<>();
+        for (String value : listSchemas) {
+            names.add(value.toLowerCase(ENGLISH));
+        }
+        return ImmutableList.copyOf(names);
+    }
+
+    @Override
+    public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
+    {
+        String schemaValue = schemaName.orElse("");
+        String dataSourceSpecificSchemaName = getDataSourceSpecificSchemaName(config, schemaValue);
+        List<String> listTables = extractSchemaAndTableData(Optional.ofNullable(dataSourceSpecificSchemaName), session);
+        List<SchemaTableName> tables = new ArrayList<>();
+        for (String value : listTables) {
+            tables.add(new SchemaTableName(dataSourceSpecificSchemaName.toLowerCase(ENGLISH), value.toLowerCase(ENGLISH)));
+        }
+
+        return tables;
+    }
+
+    public List<String> extractSchemaAndTableData(Optional<String> schema, ConnectorSession connectorSession)
+    {
+        try (ArrowFlightClient client = clientHandler.getClient(Optional.empty())) {
+            List<String> names = new ArrayList<>();
+            ArrowFlightRequest request = getArrowFlightRequest(config, schema.orElse(null));
+            ObjectNode rootNode = (ObjectNode) objectMapper.readTree(request.getCommand());
+
+            String modifiedQueryJson = objectMapper.writeValueAsString(rootNode);
+            byte[] queryJsonBytes = modifiedQueryJson.getBytes(StandardCharsets.UTF_8);
+            Iterator<Result> iterator = client.getFlightClient().doAction(new Action("discovery", queryJsonBytes), clientHandler.getCallOptions(connectorSession));
+            while (iterator.hasNext()) {
+                Result result = iterator.next();
+                String jsonResult = new String(result.getBody(), StandardCharsets.UTF_8);
+                List<String> tableNames = objectMapper.readValue(jsonResult, new TypeReference<List<String>>() {
+                });
+                names.addAll(tableNames);
+            }
+            return names;
+        }
+        catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    protected Type overrideFieldType(Field field, Type type)
+    {
+        String columnLength = field.getMetadata().get("columnLength");
+        int length = columnLength != null ? Integer.parseInt(columnLength) : 0;
+
+        String nativeType = field.getMetadata().get("columnNativeType");
+
+        if ("CHAR".equals(nativeType) || "CHARACTER".equals(nativeType)) {
+            return CharType.createCharType(length);
```

Comment:
Arrow flight returns utf8 vector for both char and varchar columns, so we are relying on additional information set in the arrow server to determine if this is char or varchar. This method can be overridden by implementing flight connectors to determine the additional information needed from server to identify the column type.